File: LeetCode/Algorithms/Medium/RotateArray/RotateArray.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length; 
        
         k = k % n;
        
        reverse(nums, 0, n-1); 
        reverse(nums, 0, k-1); 
        reverse(nums, k, n-1);
    }
    
    public void reverse(int[] nums, int start, int end) 
    {
        while(start < end)
        {
            int temp = nums[start]; 
            nums[start] = nums[end]; 
            nums[end] = temp; 
            start++;
            end--; 
        }
    }
}

// Time Complexity - O(N)
// Space Complexity - O(1), no additional space required

File: LeetCode/Algorithms/Medium/RotateArray/RotateArray.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        k %= nums.size();
        int size = nums.size();
        int i = 0;
        reverse(nums, 0, size-1);
        reverse(nums, 0, k-1);
        reverse(nums, k, size-1);
    }
    void reverse(vector<int>& nums, int start, int end){
        while(start < end){
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
};

// Time Complexity - O(N)
// Space Complexity - O(1)
